src/components/rating/rating.ts

```diff
@@ -0,0 +1,297 @@
+  }
+
+  protected handleMouseMove({ clientX }: MouseEvent) {
+    if (!this.isInteractive) {
+      return;
+    }
+
+    const value = this.calcNewValue(clientX);
+
+    if (this.hoverValue !== value) {
+      // Since mousemove spams a lot, only emit on a value change
+      this.hoverValue = value;
+      this.emitEvent('igcHover', { detail: this.hoverValue });
+    }
+  }
+
+  protected handleMouseEnter() {
+    if (this.isInteractive) {
+      this.hoverState = true;
+    }
+  }
+
+  protected handleMouseLeave() {
+    if (this.isInteractive) {
+      this.hoverState = false;
+    }
+  }
+
+  protected handleKeyDown({ key }: KeyboardEvent) {
+    if (!this.isInteractive) {
+      return;
+    }
+
+    let result = this.value;
+
+    switch (key) {
+      case 'ArrowUp':
+      case 'ArrowRight':
+        result += this.isLTR ? this.step : -this.step;
+        break;
+      case 'ArrowDown':
+      case 'ArrowLeft':
+        result -= this.isLTR ? this.step : -this.step;
+        break;
+      case 'Home':
+        result = this.step;
+        break;
+      case 'End':
+        result = this.max;
+        break;
+      default:
+        return;
+    }
+
+    // Verify new value is in bounds and emit
+    this.value = clamp(result, 0, this.max);
+
+    if (result === this.value) {
+      this.emitEvent('igcChange', { detail: this.value });
+    }
+  }
+
+  protected calcNewValue(x: number) {
+    const { width, left, right } = this.container.getBoundingClientRect();
+    const percent = this.isLTR ? (x - left) / width : (right - x) / width;
+    const value = this.round(this.max * percent + this.step / 2);
+
+    return clamp(value, this.step, this.max);
+  }
+
+  protected getPrecision(num: number) {
+    const [_, decimal] = num.toString().split('.');
+    return decimal ? decimal.length : 0;
+  }
+
+  protected round(value: number) {
+    value = Math.round(value / this.step) * this.step;
+    return Number(value.toFixed(this.getPrecision(this.step)));
+  }
+
+  /**
+   * Increments the value of the control by `n` steps multiplied by the
+   * step factor.
+   */
+  public stepUp(n = 1) {
+    this.value += this.round(n * this.step);
+  }
+
+  /**
+   * Decrements the value of the control by `n` steps multiplied by
+   * the step factor.
+   */
+  public stepDown(n = 1) {
+    this.value -= this.round(n * this.step);
+  }
+
+  protected renderSymbol(index: number) {
+    return this.symbolFormatter ? this.symbolFormatter(index) : this.symbol;
+  }
+
+  protected *renderSymbols() {
+    for (let i = 0; i < this.max; i++) {
+      yield html`
+        <span part="symbol ${this.size}"> ${this.renderSymbol(i)} </span>
+      `;
+    }
+  }
+
+  protected renderFractionWrapper(styles: { width: string }) {
+    return this.hoverPreview
+      ? html` <div
+          @click=${this.handleClick}
+          @mouseenter=${this.handleMouseEnter}
+          @mouseleave=${this.handleMouseLeave}
+          @mousemove=${this.handleMouseMove}
```

Comment:
We could conditionally attach the handlers instead of duplicating the template.
```suggestion
          @mouseenter=${this.hoverPreview ? this.handleMouseEnter : null}
          @mouseleave=${this.hoverPreview ? this.handleMouseLeave : null}
          @mousemove=${this.hoverPreview ? this.handleMouseMove : null}
```